### PR TITLE
error rewrite

### PR DESCRIPTION
**Pull request by FredTheDino**
Change the way errors are handled in the VM. This way is a lot cleaner and
does, according to me, produce better error messages.

I also accidenteally fixed a type error we have had, for like ages.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -12,9 +12,16 @@ use crate::tokenizer::Token;
...
+    TypeMismatch(Type, Type),
+    CannotInfer(Type, Type),
+    ArgumentType(Vec<Type>, Vec<Type>),
+    IndexError(Value, Type),
+
     /// (External function, parameters)
     ExternTypeMismatch(String, Vec<Type>),
-    RuntimeTypeError(Op, Vec<Value>),
+    ValueError(Op, Vec<Value>),
+    UnkownField(Value, String),
```
Comment (sornas) [inline src/error.rs:23]:
```suggestion
    UnknownField(Value, String),
```

Comment (sornas) [reply src/error.rs:23]: Followed by a `s/../..`

File: src/lib.rs [hunk 2]
```diff
@@ -730,17 +730,34 @@ mod tests {
...
+                for err in errs {
+                    eprint!(" ErrorKind::{:?} ", err.kind);
+                }
+                eprint!("]\n    WANT: [");
+                $(
+                eprint!(" {} ", stringify!($kind));
+                )*
+                eprintln!("]");
+                assert!(false);
+            }
```
Comment (sornas) [inline src/lib.rs:760]: Same comment here as in #72

File: src/vm.rs [hunk 3]
```diff
@@ -584,11 +579,13 @@ impl VM {
 
                 if let Value::BlobInstance(ty, _) = inst {
                     let ty = &self.blobs[ty].fields.get(field).unwrap().1;
-                    if ty != &Type::from(&value) {
-                        error!(self, ErrorKind::RuntimeTypeError(op, vec![inst]));
+                    let expected = Type::from(&value);
+                    if ty != &expected {
+                        error!(self, ErrorKind::TypeMismatch(expected, ty.clone()),
+                               "Field and variables type deosn't match.");
```
Comment (sornas) [inline src/vm.rs:585]:
```suggestion
                               "Types of field and variable don't match.");
```

File: src/vm.rs [hunk 4]
```diff
@@ -632,12 +638,8 @@ impl VM {
...
                         if top_type != Type::Unknown => {}
                     (a, b) if a != &b => {
-                        error!(self,
-                            ErrorKind::TypeError(
-                                op,
-                                vec![a.clone(), b.clone()]),
-                                format!("Tried to assign a type {:?} to type {:?}.", a, b)
-                        );
+                        error!(self, ErrorKind::TypeMismatch(a.clone(), b.clone()),
+                               "Cannot assign missmatching types.");
```
Comment (sornas) [inline src/vm.rs:642]:
```suggestion
                               "Cannot assign mismatching types.");
```

File: src/lib.rs [hunk 5]
```diff
@@ -730,17 +730,34 @@ mod tests {
...
     macro_rules! assert_errs {
         ($result:expr, [ $( $kind:pat ),* ]) => {
-            eprintln!("{} => {:?}", stringify!($result), $result);
-            assert!(matches!(
-                $result.unwrap_err().as_slice(),
+            let errs = if let Err(errs) = $result {
+                errs
+            } else {
+                eprintln!("    PROGRAM SUCCEEDED, WHEN IT SHOULD FAIL!");
+                unreachable!();
```
Comment (sornas) [inline src/lib.rs:737]: Still shouty.

(But not in #72)

File: src/lib.rs [hunk 6]
```diff
@@ -730,17 +730,34 @@ mod tests {
...
+            };
```
Comment (sornas) [inline src/lib.rs:738]: .

File: src/lib.rs [hunk 7]
```diff
@@ -730,17 +730,34 @@ mod tests {
     #[macro_export]
     macro_rules! assert_errs {
         ($result:expr, [ $( $kind:pat ),* ]) => {
-            eprintln!("{} => {:?}", stringify!($result), $result);
-            assert!(matches!(
-                $result.unwrap_err().as_slice(),
+            let errs = if let Err(errs) = $result {
+                errs
+            } else {
+                eprintln!("    PROGRAM SUCCEEDED, WHEN IT SHOULD FAIL!");
```
Comment (FredTheDino) [inline src/lib.rs:736]:
```suggestion
                eprintln!("    Program succeeded, when it should fail");
```

**Review by sornas — APPROVED** [review]

**Pull request merged**